### ISSUE TITLE
aux_loss does not update any weigth

### ISSUE THREAD

**Issue by steb6** [lucidrains/compressive-transformer-pytorch#5]
Hi lucidrains,
thanks for your implementation, it is very elegant and helped me a lot with my disertation.
Anyway I can't understand a particular: it seems like aux_loss is not related to any weight because of the detaching in the last part of the SelfAttention layer. With the following code, for example, I get that there is no layer optimized by aux_loss:

```
import torch
from compressive_transformer_pytorch import CompressiveTransformer
from compressive_transformer_pytorch import AutoregressiveWrapper

model = CompressiveTransformer(
    num_tokens = 20000,
    dim = 512,
    depth = 6,
    seq_len = 1024,
    mem_len = 1024,
    cmem_len = 256,
    cmem_ratio = 4,
    memory_layers = [5,6]
).cuda()

model = AutoregressiveWrapper(model)

inputs = torch.randint(0, 20000, (1, 1024)).cuda()

optimizer = torch.optim.Adam(model.parameters())

for loss, aux_loss, _ in model(inputs, return_loss = True):
    optimizer.zero_grad(set_to_none=True)
    loss.backward(retain_graph=True)
    print("OPTIMIZED BY LOSS ************************************************************")
    for module_name, parameter in model.named_parameters():
        if parameter.grad is not None:
            print(module_name)
    optimizer.zero_grad(set_to_none=True)
    aux_loss.backward(retain_graph=True)
    print("OPTIMIZED BY AUX_LOSS ************************************************************")
    for module_name, parameter in model.named_parameters():
        if parameter.grad is not None:
            print(module_name)
```
I am not expert about the PyTorch mechanisms, so maybe I am getting something wrong.
Again thank you

**Comment by vpj** [comment]
I guess the problem is https://github.com/lucidrains/compressive-transformer-pytorch/blob/a73d146ee7bc9e47cc6c525f645a5bbb583537ad/compressive_transformer_pytorch/compressive_transformer_pytorch.py#L278

The paper states that the key, value, query projection weights should be detached. Not the key, value, query projects

**Comment by lucidrains** [comment]
![Screenshot from 2021-01-31 08-41-33](https://user-images.githubusercontent.com/108653/106391030-27e26e00-63a0-11eb-9781-7578ce61dd40.png)

@StefanoBerti omg, thank you for catching this! I have corrected it so that the convolutional compression net is updated properly :pray:

**Comment by vpj** [comment]
hi @lucidrains

wouldn't https://github.com/lucidrains/compressive-transformer-pytorch/blob/d34aab11398a399896e1f9014ffb5db431eb16b8/compressive_transformer_pytorch/compressive_transformer_pytorch.py#L271 detach it permanently?
